src/lib/optimizer/join_ordering/join_graph_builder.cpp

```diff
@@ -0,0 +1,168 @@
+#include "join_graph_builder.hpp"
+
+#include "logical_query_plan/join_node.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/union_node.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+std::shared_ptr<JoinGraph> JoinGraphBuilder::operator()(const std::shared_ptr<AbstractLQPNode>& lqp) {
+  /**
+   * Traverse the LQP until the first non-vertex type (e.g. a UnionNode) is found or a node doesn't have precisely
+   * one child. This way, we traverse past Sort/Aggregate etc. nodes that later form the "parents" of the JoinGraph
+   */
+  auto current_node = lqp;
+  while (_lqp_node_type_is_vertex(current_node->type())) {
+    if (!current_node->left_child() || current_node->right_child()) {
+      break;
+    }
+
+    current_node = current_node->left_child();
+  }
+  const auto parent_relations = current_node->parent_relations();
+
+  // Traverse the LQP, identifying JoinPlanPredicates and Vertices
+  _traverse(current_node);
+
+  return JoinGraph::from_predicates(std::move(_vertices), std::move(parent_relations), _predicates);
+}
+
+void JoinGraphBuilder::_traverse(const std::shared_ptr<AbstractLQPNode>& node) {
+  // Makes it possible to call _traverse() on children without checking whether they exist first.
+  if (!node) {
+    return;
+  }
+
+  if (_lqp_node_type_is_vertex(node->type())) {
+    _vertices.emplace_back(node);
+    return;
+  }
+
+  switch (node->type()) {
+    case LQPNodeType::Join: {
+      /**
+       * For now, only inner joins are turned into predicates for the JoinGraph. Cross joins are simply being traversed
+       * past and all other joins are turned into vertices.
```

Comment:
Could you clarify the difference between predicates and vertices a bit?

Reply:
The JoinGraph is now doc'ed as
```
 * Represents a connected subgraph of an LQP, with the LQPNodes "above" contained in the parent_relations and the
 * subplans "below" it being the vertices.
 * The JoinGraph clusters Predicates operating on the same set of vertices into JoinEdges.
```

is that sufficient?

Reply:
Let me try it in a different way: I understand the two lines that I attached this comment to as follows:

Inner joins become predicates. Predicates (i.e., edges) link multiple tables (i.e, vertices).

In my understanding, this conflicts with "other joins are turned into vertices". How can a join be either an edge or a vertex, depending on if it is an inner or an outer join?

Reply:
The Predicate "contained" in an Inner Join can be turned into a JoinPlanPredicate and traversal can continue in its children. A semi join's predicate can't be turned into a JoinPlanPredicate and thus traversal has to stop and the "SemiJoinNode" becomes a vertex. 
It's all about whether nodes contain Predicates that can be turned into JoinPlanPredicates.

Reply:
Covering an outer join in a JoinOrdering algorithm is HARD. There is a paper, but I'm not sure whether I'll ever include them in the JoinOrdering process.

Reply:
Ok, I start to understand. Now I also see that in the case of an outer join, you stop the traversal. That should be part of the comment:

"Cross joins are simply being traversed past.
Outer joins cannot be described in a JoinGraph. The table created by the outer join is added as a vertex and traversal stops at this point."

Reply:
If that clarifies it, please tell me which part to include in the docs, to me this is blindingly obvious because I am massively betriebsblind here

Reply:
A short comment that you decided not to do this for now would be great. Otherwise, people might wonder if it is a fundamental limitation.

Reply:
* Cross joins are simply being traversed past. Outer joins are hard to address during JoinOrdering and until we 
       * do, outer join predicates are not included in the JoinGraph.
       * The outer join node is added as a vertex and traversal stops at this point.

Reply:
Perfect.

Reply:
Don't want to talk about "The table created by the outer join" here. Tables are a concept alien to the abstraction level of the JoinGraph/JoinOrdering